### ISSUE TITLE
Must use accents when searching by author/reviewer name in editorial backend

### ISSUE THREAD

**Issue by NateWr** [pkp/pkp-lib#7140]
**Describe the problem you would like to solve**
When searching for a submission by the author's name, or a reviewer during review assignment, searches won't match letters with diacritics/accents. For example, a search for `edicao` will not return a submission with an author named `edição`.

This can make it difficult for editorial staff working with multilingual journals, who may not be used to typing diacritics/accents. Or even in monolingual journals where authors or reviewers have used the correct accented characters when entering their name.

**Describe the solution you'd like**
Where possible, search phrases should be matched against accented characters, so that a search for `edicao` will return `edição`.

**Who is asking for this feature?**
This was reported in a free text field in the TC community survey.

**Comment by NateWr** [comment]
Ok, thanks to @ctgraham, I've learned that MySQL supports this if you are using the correct collation (maybe `utf8_unicode_ci` or `utf8_general_ci`). But Postgresql requires the use of an `unaccent()` function.  Some resources:

- https://dba.stackexchange.com/questions/156907/equivalent-of-utf8-unicode-ci-collation-in-postgresql
- https://stackoverflow.com/a/4813651/1723499

**Comment by asmecher** [comment]
If it looks like supporting both databases natively will get hairy, another approach would be to have a "search name" field/setting row in the database. This would be the various searchable fields concatenated, lowercased, and stripped of accents; then, rather than
```
... WHERE LOWER(UNACCENT(family_name_setting.setting_value)) LIKE '%smith%' 
OR UNACCENT(given_name_setting.setting_value) LIKE '%Smith%' 
OR UNACCENT(display_name_setting.setting_value) LIKE '%Smith%' 
```
...it would be possible to use just...
```
... WHERE search_name_setting.setting_value LIKE '%smith%' 
```
...where `family_name_setting` is a row in `user_settings` that gets populated with `strtolower(iconv(...))` on insert/update. Or, perhaps even better, it could even be an indexed column in `users`, since searching `user_settings` involves a full table scan.

(All of this can go away when we can count on good support for derived columns and functional indexes in our base versions of both MySQL and PostgreSQL, but that appears not to be the case yet.)

**Comment by ctgraham** [comment]
>  various searchable fields concatenated, lowercased, and stripped of accents

Sounds like a lot of internal effort to re-create a wheel named "Solr".

**Comment by asmecher** [comment]
Yes, but almost none of our community will be able to run that :(

We used to have a similar field called "cleanTitle" for title-based sorting of articles without leading quotes etc; it was pretty low-fuss to maintain.

**Comment by jonasraoni** [comment]
There are some search engines in PHP, I even found what looks like a [PHP implementation of Lucene](https://github.com/pucene). I doubt they will have a great performance, if I'm wrong, it's a possibility to consider =]

If we're going to introduce custom queries, I think a full-text index on the column with the concatenated content will have a better outcome 🤔

Personally I'm curious about the coverage of removing accents against other peculiarities that may exist in other languages, and if/how these peculiarities are handled by the search engines (after searching fast, looks like the support for such things is limited and requires custom plugins/settings 👀).

**Comment by diegoabadan** [comment]
> There are some search engines in PHP, I even found what looks like a PHP implementation of Lucene. I doubt they will have a great performance, if I'm wrong, it's a possibility to consider =]

Unfortunately, it looks abandoned.

**Comment by GaziYucel** [comment]
I would go for the "search_text" field which holds a concatenated search text as asmecher suggests. I would fill this field through triggers in the database or a scheduled task in PHP. I would store the original string and the stripped string (all lower case) if these are different. Advantage of this is that the search query can be relatively non-special, fast and easy to implement and also searchable by the original text.

**Comment by NateWr** [comment]
If this is a solved problem in MySQL (through collation), we could simply consider that our recommended database. It's not ideal, but I am wary of trying to architect a comprehensive solution here. We would eventually have expectations to roll this out to every entity that is searched:

- Submissions
- Users
- Issues
- Announcements
- Contexts
- Email Templates

And as we develop new features, additional entities will attract similar search requirements:

- Activity logs
- Discussions
- ...

We may be able to bake this into our `EntityDAO` classes to some extent, to make it easy to define and use these indexes. But just be aware that if we roll our own approach it's going to be a lot more than just submissions and users.

**Comment by ctgraham** [comment]
> If this is a solved problem in MySQL (through collation), we could simply consider that our recommended database.

Seconded.  This fits with the general principle of, "if we're not adding something novel to the solution space, we should first look to reuse other's existing work".

**Comment by asmecher** [comment]
Do you mean that we treat the problem as solved using MySQL (since it's built-in), and recommend the use of MySQL for those that need this feature? I'm OK with that, I think. We may need to solve the problem of searching performance in BLOB `setting_value` fields, but that's separate.

**Comment by asmecher** [comment]
Closing as a WONTFIX -- it's solved in MySQL via collations, and not worth the trouble of implementing for PostgreSQL users alone.
